### PR TITLE
tweaks to enable Sona with Zyla backend support

### PR DESCRIPTION
**Pull request by csoeller**
Addresses issue #425.

**Is this a bugfix or an enhancement?**

Enhancement of the Zyla backend to allow Sona heads to be connected.

**Proposed changes:**

- Changes to the AndorZyla code to self-detect some capabilities.

- Additional interface changes in ZylaControlPanel so that the cooling mode options of the Sona are autodetected and offered in the control panel as options for selection.

**Checklist:**

- [x] Tested with numpy=1.16
- [x] Tested on python 2.7 only
- [NA ] Tested with wx=3.x and wx=4.x [if UI code]
- [x] Does the PR avoid variable renaming in existing code, whitespace changes, and other forms of tidying? [There is a place for code tidying, but it makes reviewing 
much simpler if this is kept separate from functional changes]

If an enhancement (or non-trivial bugfix):

- [x] Has this been discussed in advance (feature request, PR proposal, email, or direct conversation)? [has been raised]
- [-] Does this change how users interact with the software? How will these changes be communicated?
- [x] Does this maintain backwards compatibility with old data?
- [-] Does this change the required dependencies?
- [-] Are there any other side effects of the change? [don't think so but will discuss]

### CODE REVIEW

File: PYME/Acquire/Hardware/AndorNeo/AndorZyla.py [hunk 1]
```diff
@@ -130,26 +143,9 @@ def noise_properties(self):
...
-        def getAvailableValues(self):
-            return self.gainmodes
-
-        def setString(self,str):
-            self.cam.SetSimpleGainMode(str)
-
-        def getString(self):
-            return self.cam.GetSimpleGainMode()
-
-
```
Comment (csoeller) [inline PYME/Acquire/Hardware/AndorNeo/AndorZyla.py:150]: The SimpleGainEnumClass seemed to be essentially overhead only once the code was investigated. I got rid of it in favour of a simple keyword for EnumControl objects in the Zyla Control Panel.

File: PYME/Acquire/Hardware/AndorNeo/AndorZyla.py [hunk 2]
```diff
@@ -653,8 +677,13 @@ def GenStartMetadata(self, mdh):
...
             mdh.setEntry('Camera.ROIWidth', x2 - x1)
             mdh.setEntry('Camera.ROIHeight', y2 - y1)
-            #mdh.setEntry('Camera.StartCCDTemp',  self.GetCCDTemp())
 
+            #mdh.setEntry('Camera.StartCCDTemp',  self.GetCCDTemp())
+            # these make sense with the Sona choices but should also be ok for the Zyla
+            mdh.setEntry('Camera.TemperatureControl', self.TemperatureControl.getString())
+            mdh.setEntry('Camera.TemperatureStatus', self.TemperatureStatus.getString())
+            mdh.setEntry('Camera.SensorTemperature', self.SensorTemperature.getValue())
+
```
Comment (csoeller) [inline PYME/Acquire/Hardware/AndorNeo/AndorZyla.py:712]: With the Sona temperature control options it becomes important to know in which cooling mode the camera was operated as I expect this to affect camera maps. It cannot hurt for the Zyla either.

File: PYME/Acquire/Hardware/AndorNeo/ZylaControlPanel.py [hunk 3]
```diff
@@ -8,14 +8,24 @@
...
+        # this keyword allows us to override unwieldy property Names
+        if targetPropertyName is None:
+            self.targetPropertyName = target.propertyName
+        else:
+            self.targetPropertyName = targetPropertyName
+            
+        try:
+            self.targetReadOnly = target.isReadOnly()
+        except AttributeError:
+            self.targetReadOnly = False
```
Comment (csoeller) [inline PYME/Acquire/Hardware/AndorNeo/ZylaControlPanel.py]: The ReadOnly check is a kludge to support the read only temperature status property in the ControlPanel. It should really be a status display of some form. This probably needs discussion.

File: PYME/Acquire/Hardware/AndorNeo/ZylaControlPanel.py [hunk 4]
```diff
@@ -73,16 +84,20 @@ def __init__(self, parent, cam, scope):
         self.cam = cam
         self.scope = scope
         
-        self.ctrls = [EnumControl(self, cam.SimpleGainEnumInstance),
+        self.ctrls = [EnumControl(self, cam.SimplePreAmpGainControl, targetPropertyName='SimpleGainMode'),
```
Comment (csoeller) [inline PYME/Acquire/Hardware/AndorNeo/ZylaControlPanel.py]: here we got rid of the SimpleGainEnumInstance which really only provided a simpler name for the property. This has insteads been done with a simpler keyword to the EnumControl constructor.

File: PYME/Acquire/Hardware/AndorNeo/ZylaControlPanel.py [hunk 5]
```diff
@@ -73,16 +84,20 @@ def __init__(self, parent, cam, scope):
...
                       BoolControl(self, cam.SpuriousNoiseFilter),
                       BoolControl(self, cam.StaticBlemishCorrection),
-                      EnumControl(self, cam.CycleMode),]
-        
+                      EnumControl(self, cam.CycleMode),
+                      EnumControl(self, cam.TemperatureStatus),]
+
+        if len(cam.TemperatureControl.getAvailableValues()) > 0: # need to check that Zyla has at least one mode
+            self.ctrls.append(EnumControl(self, cam.TemperatureControl))
+
```
Comment (csoeller) [inline PYME/Acquire/Hardware/AndorNeo/ZylaControlPanel.py:98]: Here we could check for >1 instead which would avoid the Zyla case where only one mode is available (0.0 deg). So far I opted for the verbose one that just shows the single selectable option.

File: PYME/Acquire/Hardware/AndorNeo/AndorZyla.py [hunk 6]
```diff
@@ -74,6 +74,19 @@ class AndorBase(SDK3Camera, CameraMapMixin):
...
+                'ElectronsPerCount' : 0.45,
+                'ADOffset' : 100, # check mean (or median) offset
+                'SaturationThreshold' : 1776  #(2**16 -1) # check this is really 11 bit
+            },
+            u'16-bit (high dynamic range)': {
+                'ReadNoise' : 1.84,
+                'ElectronsPerCount' : 1.08,
+                'ADOffset' : 100,
+                'SaturationThreshold' : 44185
+            }},
```
Comment (csoeller) [inline PYME/Acquire/Hardware/AndorNeo/AndorZyla.py:89]: This showcases the difference for the Sona which has a different set of gain modes, only two are supported and the 16bit version has a slightly different name to the Zyla.

As we discussed, eventually this info should go in a config file but this should be the subject of a separate PR/issue as it is essentially orthogonal to the Sona functionality tweaks and also needs a bit more discussion.

File: PYME/Acquire/Hardware/AndorNeo/AndorZyla.py (hunk 1 above)
Comment (David-Baddeley) [reply PYME/Acquire/Hardware/AndorNeo/AndorZyla.py:150]: I recall there being a good reason for this (which might have evaporated with a more recent SDK3 version), but can't put my finger on it. It could be related to Andor Neo support.

Comment (David-Baddeley) [reply PYME/Acquire/Hardware/AndorNeo/AndorZyla.py:150]: This is needed - and is not just an alias for `SimplePreAmpGainControl`, as `SetSimpleGainMode()` actually sets two properties (the gain mode **and** pixel encoding which is not set automatically by setting the `SimplePreAmpGainControl` property).

File: PYME/Acquire/Hardware/AndorNeo/ZylaControlPanel.py [hunk 7]
```diff
@@ -73,16 +84,20 @@ def __init__(self, parent, cam, scope):
...
         
-        self.ctrls = [EnumControl(self, cam.SimpleGainEnumInstance),
+        self.ctrls = [EnumControl(self, cam.SimplePreAmpGainControl, targetPropertyName='SimpleGainMode'),
                       EnumControl(self, cam.PixelReadoutRate),
                       BoolControl(self, cam.SpuriousNoiseFilter),
                       BoolControl(self, cam.StaticBlemishCorrection),
-                      EnumControl(self, cam.CycleMode),]
-        
+                      EnumControl(self, cam.CycleMode),
+                      EnumControl(self, cam.TemperatureStatus),]
```
Comment (David-Baddeley) [inline PYME/Acquire/Hardware/AndorNeo/ZylaControlPanel.py]: Temperature status would probably be better reported through the status bar - e.g. either by adding a callback to `microscope.StatusCallbacks`, or by re-working the exisiting camera temperature stuff in `microscope.genStatus()` to defer the status stuff a bit more to the camera / let it decide what to report for temperature etc ...

There might also be either performance or synchronisation issues in polling the temperature and/or temperature status through an Enum control. As it stands, the Zyla code has the temperature being queried only when`StartExposure()` is called, with the result cached and the cached value used for subsequent calls to `GetCCDTemp()`. It's been a while, but I'm pretty sure this was done this way on purpose to address a performance problem.

Comment (csoeller) [reply PYME/Acquire/Hardware/AndorNeo/ZylaControlPanel.py]: Now addressed as a new method that can be added to `microscope.StatusCallbacks` in init file. The method is called `TemperatureStatusText` and I add below a screenshot that shows it in action. With the current SDK no performance issues apparently.

pyme-zyla-sona-tweaks

File: PYME/Acquire/Hardware/AndorNeo/AndorZyla.py (hunk 1 above)
Comment (csoeller) [reply PYME/Acquire/Hardware/AndorNeo/AndorZyla.py:150]: I reinstated SimplePreampGainControl and see the reason for the pixel encoding. Interestingly, with the latest SDK there were no issues not doing it although I can't be sure it was stuck on 16bit. Anyway, seems to work fine now with latest commits.

File: PYME/Acquire/Hardware/AndorNeo/ZylaControlPanel.py (hunk 5 above)
Comment (csoeller) [reply PYME/Acquire/Hardware/AndorNeo/ZylaControlPanel.py:98]: Changed to >1 so only present if there is a real choice.
